File: src/c++/Golomb.cpp
#include <stdio.h>
#include <iostream>
#include <math.h>
#include <vector>
#include <algorithm>
#include "BitStream.cpp"
#include <chrono>

using namespace std;

class Golomb
{
private:
    int encoding_parameter;
    int b_param;
    int unary_limit;
    vector<bool> bit_feed;

public:
    /* Constructor of the Golomb converter */
    Golomb(int m)
    {
        encoding_parameter = m;
        b_param = ceil(log2(m));

        unary_limit = pow(2, b_param) - encoding_parameter; //Encode the first 2b−m values of r using the first 2b−m binary codewords of b−1 bits
    }

    int enconding_parameter()
    {
        return encoding_parameter;
    }

    vector<bool> encode(int m)
    {

        bool sign = m < 0;
        m = abs(m);

        int quotient = m / encoding_parameter,
            remainder = m % encoding_parameter;

        vector<bool> code;
        int i;

        // Unary code
        code.push_back(sign);
        for (i = 1; i < quotient + 1; i++)
        {
            code.push_back(true);
        }
        code.push_back(false); //Start binary code

        // Binary Code
        // Coversion of remainder to binary
        for (int j = b_param - 1; j >= 0; j--, i++)
        {
            int k = remainder >> j;
            code.push_back(k & 1);
        }

        return code;
    }

    bool can_decode() {
        if (bit_feed.size() < 1)
            return false;
        vector<bool>::iterator it = find(bit_feed.begin() + 1, bit_feed.end(), false);
        if (it == bit_feed.end())
            return false;
        return distance(it, bit_feed.end()) > b_param;
    }

    bool add_bits(vector<bool> bits) {
        bit_feed.insert(bit_feed.end(), bits.begin(), bits.end());
        return can_decode();
    }

    void decode_nums(vector<int> *num_array){
        while (can_decode())
            num_array->push_back(decode());
    }

    int decode()
    {
        int value, index;

        bool sign = bit_feed.at(0);

        vector<bool>::iterator it = find(bit_feed.begin() + 1, bit_feed.end(), 0);
        value = distance(bit_feed.begin() +1 , it++) * encoding_parameter;
    
        for (int j = b_param-1; j >= 0; j--, it++)
            value += *it << j;

        sign ? value = value *-1 : value = value;

        bit_feed.erase(bit_feed.begin(), it);

        return value;
    }


    bool add_bit(bool bit)
    {
        bit_feed.push_back(bit);
        return can_decode();
    }

    vector<int> decode_nums()
    {
        vector<int> num_array;

        while (can_decode())
            num_array.push_back(decode());
        return num_array;
    }

    int decode(vector<bool> encoded)
    {
        bool sign = encoded.at(0);

        int quotient, remainder, index;
        quotient = remainder = 0;
        index = 1;
        while (1)
        {
            bool bit = encoded.at(index++);
            if (!bit)
                break;
            quotient++;
        }

        for (int j = b_param; j > 0; j--)
        {
            bool bit = encoded.at(index++);
            remainder += bit << j - 1;
        }

        int value = quotient * encoding_parameter + remainder;
        sign ? value = value *-1 : value = value;
        encoded.erase(encoded.begin(), encoded.begin() + index);

        return value;
    }
};

File: src/python/Golomb.py
import math
import BitStream

class Golomb:
    def __init__(self,m):
        self.encoding_parameter = m
        self.b_param = math.ceil(math.log2(m))

        self.unary_limit = pow(2, self.b_param) - self.encoding_parameter; #Encode the first 2b−m values of r using the first 2b−m binary codewords of b−1 bits
    
        self.bit_feed = []

    def encode(self, value):
        code = [True, value<0]
        value = abs(value)

        quotient = value // self.encoding_parameter
        remainder = value % self.encoding_parameter

        code = code + [True] * quotient + [False]   #Encode the quotient - UNARY
        
        for i in range(self.b_param-1, -1, -1):
            code.append(bool((remainder >> i) & 1))
        return code

    def can_decode(self):
        """
        if self.bit_feed == []:
            return False
        
        bit_test = self.bit_feed[1:]
        
        while True:
            if bit_test == []:
                return False
            bit = bit_test.pop(0)
            if not bit:
                break

        return len(bit_test) >= (self.b_param)
        """
        if len(self.bit_feed) < 2:
            return False
        
        bit_test = self.bit_feed[2:]
        
        while True:
            if bit_test == []:
                return False
            bit = bit_test.pop(0)
            if not bit:
                break

        return len(bit_test) >= (self.b_param)


    def add_bits(self, bits):
        self.bit_feed += bits
        return self.can_decode()

    def add_bit(self, bit):
        self.bit_feed.append(bit)
        return self.can_decode()
    
    def decode_nums(self):
        num_array = []
        while self.can_decode():
            if self.bit_feed[0]:
                num_array.append(self.decode(self.bit_feed))
            else:
                self.bit_feed.pop(0)

        return num_array

    def decode(self, code):
        code.pop(0)
        sign_bit = code.pop(0)

        quotient = 0
        remainder = 0

        while True:
            bit = code.pop(0)
            if not bit:
                break
            quotient+=1

        for i in range(self.b_param-1, -1, -1):
            bit = code.pop(0)
            remainder += int(bit) << i

        value = quotient*self.encoding_parameter + remainder
        if(sign_bit):
            value = -value

        return value

def main():
    gomby = Golomb(4) #Kawaii desu-nee?
    bity = BitStream.BitStream()
    
    for i in range(-16, 0):
        testis = gomby.encode(i)
        bity.add_to_bit_array(testis)

    bity.write_allbits("test.bin")

    bity.reset_bit_array()
    bity.clear_padding()
    
    for i in range (16):
        testis = gomby.encode(i)
        bity.add_to_bit_array(testis)
    bity.write_allbits("test.bin")

    bity2 = BitStream.BitStream("test.bin")
    array_of_nums = []
    number_of_numbers = 0
    while bity2.read_bits(20):
        got_number = gomby.add_bits(bity2.get_bit_array())
        bity2.delete_bits(20)

        if got_number:
            nums = gomby.decode_nums()
            array_of_nums += nums
            number_of_numbers += len(nums)
        
    bity2.read_allbits()
    got_number = gomby.add_bits(bity2.get_bit_array())
    nums = gomby.decode_nums()
    array_of_nums += nums
    print(array_of_nums)

if __name__ == "__main__":
    main()
